packages/amplify-graphql-auth-transformer/src/vtl-generator/ddb/resolvers/helpers.ts

```diff
@@ -165,12 +166,31 @@ export const iamExpression = (
         iff(not(ref(IS_AUTHORIZED_FLAG)), iamCheck(role.claim!, set(ref(IS_AUTHORIZED_FLAG), bool(true)), hasIdentityPoolId)),
       );
     });
-  } else {
+  }
+
+  if (genericIamAccessEnabled) {
+    expression.push(genericIamAccessExpression());
+  } else if (roles.length === 0) {
     expression.push(ref('util.unauthorized()'));
   }
+
   return iff(equals(ref('util.authType()'), str(IAM_AUTH_TYPE)), compoundExpression(expression));
 };
 
+/**
+ * Creates an expression that allows generic IAM access for principals not associated to CognitoIdentityPool
+ */
+export const genericIamAccessExpression = (): Expression => {
+  const doesNotHaveIdentityPoolExpression = and([
```

Comment:
Thinking through this--I think `and` is the right expression to use here. If either identityId or identityPoolId is missing, the identity pool evaluation should fail, but we don't want to automatically fall back to IAM authorization if it's in an inconsistent state like that. 

We should verify that there are no cases where one is present without the other, but I think this is right.

Reply:
Sounds good. I'll add something to handle inconsistent state - likely a validation that fails request (unauthorized seems to be wrong error for this case).

Reply:
I ended up reversing this check and making DDB generators look like RDS.
Not sure if consistency check is needed now.